src/systemd-sonic-generator/systemd-sonic-generator.c

```diff
@@ -162,30 +166,31 @@ static void replace_multi_inst_dep(char *src) {
     fp_tmp = fopen(tmp_file_path, "w");
 
     while ((nread = getline(&line, &len, fp_src)) != -1 ) {
-        if (strstr(line, "[Service]") != NULL) {
+        if ((strstr(line, "[Service]") != NULL) || 
+            (strstr(line, "[Timer]") != NULL)) {
             section_done = true;
             fputs(line,fp_tmp);
-        } else if (strstr(line, "[Install]") != NULL)  {
+        } else if (strstr(line, "[Install]") != NULL) {
             section_done = false;
             fputs(line,fp_tmp);
         } else if ((strstr(line, "[Unit]") != NULL) ||
            (strstr(line, "Description") != NULL) ||
-           (section_done == true)){
+           (section_done == true)) {
             fputs(line,fp_tmp);
         } else {
             line_copy = strdup(line);
-            token = strtok(line_copy, "=");
-            while ((word = strtok(NULL, " "))){
+            token = strtok_r(line_copy, "=", &save_ptr1);
```

Comment:
>strtok_r [](start = 20, length = 8)

Rewrite everything by python?

Reply:
The initial change to add systemd-generator had a discussion on this after which it was decided that C will be used. Guohan had pointed to a reference: "If you are careful, you can implement generators in shell scripts. We do recommend C code however, since generators are executed synchronously and hence delay the entire boot if they are slow." https://www.freedesktop.org/software/systemd/man/systemd.generator.html.

Reply:
Thanks for sharing the history discussion.
1. I don't understand why it is on boot critical path. Could you elaborate? It seems relating the ASIC numbers, and only need to run once on first cold boot.

2. Can we remove explicit knowledge on ASIC numbers, just like depending on all syncd@N.service ?

3. Can we use C++ (as high performance as C) string and stream to split strings?

---
In reply to: [443869780](https://github.com/Azure/sonic-buildimage/pull/4820#discussion_r443869780) [](ancestors = 443869780)

Reply:
1. systemd-generator runs before systemd  starts, hence it is on boot critical path.
Yes, we will ideally need to run this only once on first cold boot. Currently this is always run upon each boot before systemd starts.
2. systemd generator does 2 things based on number of asics: 1. create the number of instances to be started, ex: swss@0, swss@1 etc. 2. Change the dependency of host services to reflect this list of actual running services. if we specify syncd@N - systemd does not convert that to 0 - n-1, instead looks for syncd@N service, which is not running.
3. Agreed that c++ will provide in built functions for string, but as service files are going to have some set of strings which will not vary much, c string functions should satisfy that requirement. Was there any other reason to use c++ other than ease of use with string functions?

Reply:
Did a comparison on runtime between c and c++. On a mulit-asic platform, C program takes 2643 micro seconds and c++ take 2651 micro seconds. As there is not much difference, we can move to c++. Will raise another PR to move to code c++ with stream functions for string operations to make it less error prone.